File: crates/powder-java/java/com/powder/Orm.java
package com.powder;

import java.util.List;
import java.util.Map;

/**
 * The model layer over a {@link Client}: the same operation semantics as the
 * TS/Python/Go/C# ORMs, executed by the shared Rust engine. Options are maps
 * with the unified keys — {@code where}, {@code orderBy}, {@code limit},
 * {@code offset}, {@code include}, {@code join}. Rows come back as
 * {@code List<Map<String, Object>>} (insertion-ordered maps).
 *
 * <pre>
 * try (Orm orm = db.orm(schemaJson)) {              // powder.schema.json text
 *     Orm.Table users = orm.table("users");
 *     users.create(Map.of("id", 1, "name", "alice", "score", 9.5, "active", true));
 *     List&lt;Map&lt;String, Object&gt;&gt; rows = users.findMany(Map.of(
 *         "where", Map.of("active", true, "score", Map.of("gte", 5)),
 *         "orderBy", Map.of("score", "desc"),
 *         "limit", 10));
 * }
 * </pre>
 *
 * Note: {@link Map#of} does not preserve key order — for a multi-column
 * {@code orderBy}, pass a {@link java.util.LinkedHashMap}.
 */
public final class Orm implements AutoCloseable {
    private final Client client;
    private long schema;

    Orm(Client client, String schemaJson) {
        this.client = client;
        this.schema = PowderNative.ormSchemaNew(schemaJson);
    }

    /** Handle for one table's CRUD surface. */
    public Table table(String name) {
        return new Table(name);
    }

    @Override
    public void close() {
        if (schema != 0) {
            PowderNative.ormSchemaFree(schema);
            schema = 0;
        }
    }

    private long schemaHandle() {
        if (schema == 0) {
            throw new IllegalStateException("orm is closed");
        }
        return schema;
    }

    /** One table's unified CRUD surface. */
    public final class Table {
        private final String name;

        private Table(String name) {
            this.name = name;
        }

        /** Rows matching {@code opts}; null opts = all rows. */
        @SuppressWarnings("unchecked")
        public List<Map<String, Object>> findMany(Map<String, ?> opts) {
            return (List<Map<String, Object>>) find("findMany", opts);
        }

        /** First matching row, or null. */
        @SuppressWarnings("unchecked")
        public Map<String, Object> findFirst(Map<String, ?> opts) {
            return (Map<String, Object>) find("findFirst", opts);
        }

        /** Every row. */
        public List<Map<String, Object>> all() {
            return findMany(null);
        }

        /** INSERT one row; missing (nullable) columns are omitted. */
        public long create(Map<String, ?> data) {
            return execute("create", Map.of("data", data));
        }

        /** Bulk INSERT (chunked multi-row VALUES); every row must carry the
         *  same columns as the first. */
        public long createMany(List<? extends Map<String, ?>> rows) {
            return execute("createMany", Map.of("rows", rows));
        }

        /** UPDATE matching rows; returns the affected count. */
        public long update(Map<String, ?> where, Map<String, ?> data) {
            return execute("update", Map.of("where", where, "data", data));
        }

        /** DELETE matching rows. An empty where is rejected — use
         *  {@link #deleteAll}. */
        public long delete(Map<String, ?> where) {
            return execute("delete", Map.of("where", where));
        }

        /** DELETE every row (explicit opt-in). */
        public long deleteAll() {
            return execute("deleteAll", null);
        }

        /** COUNT rows matching where (null counts everything). */
        public long count(Map<String, ?> where) {
            return execute("count", where == null ? null : Map.of("where", where));
        }

        /** Whether at least one row matches. */
        public boolean exists(Map<String, ?> where) {
            return findFirst(where == null ? Map.of("limit", 1) : Map.of("where", where, "limit", 1)) != null;
        }

        /** SUM/AVG/MIN/MAX over one column; null when no rows match. */
        public Double aggregate(String fn, String column, Map<String, ?> where) {
            Object v = find("aggregate", where == null
                    ? Map.of("fn", fn, "column", column)
                    : Map.of("fn", fn, "column", column, "where", where));
            return v == null ? null : ((Number) v).doubleValue();
        }

        /** GROUP BY with aggregates ({@code by}, {@code count}, {@code sum},
         *  {@code avg}, {@code min}, {@code max}, {@code having},
         *  {@code orderBy}, ...); aggregates come back aliased {@code _count},
         *  {@code _sum_&lt;col&gt;}, .... */
        @SuppressWarnings("unchecked")
        public List<Map<String, Object>> groupBy(Map<String, ?> opts) {
            return (List<Map<String, Object>>) find("groupBy", opts);
        }

        private long execute(String op, Map<String, ?> opts) {
            return PowderNative.ormExecute(client.handle(), schemaHandle(), opJson(op, opts));
        }

        private Object find(String op, Map<String, ?> opts) {
            String json = PowderNative.ormFindJson(client.handle(), schemaHandle(), opJson(op, opts));
            return Json.read(json);
        }

        private String opJson(String op, Map<String, ?> opts) {
            java.util.LinkedHashMap<String, Object> node = new java.util.LinkedHashMap<>();
            node.put("op", op);
            node.put("table", name);
            if (opts != null) {
                node.putAll(opts);
            }
            return Json.write(node);
        }
    }
}

File: bindings/csharp/Powder/Orm.cs
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Powder;

/// <summary>
/// The model layer over a <see cref="Client"/>: the same operation semantics
/// as the TS/Python/Go ORMs, executed by the shared Rust engine. Options are
/// anonymous objects (or dictionaries) with the unified keys — <c>where</c>,
/// <c>orderBy</c>, <c>limit</c>, <c>offset</c>, <c>include</c>, <c>join</c>.
///
/// <code>
/// using var orm = db.Orm(schemaJson);           // powder.schema.json text
/// var users = orm.Table("users");
/// users.Create(new { id = 1, name = "alice", score = 9.5, active = true });
/// var rows = users.FindMany(new
/// {
///     where = new { active = true, score = new { gte = 5 } },
///     orderBy = new { score = "desc" },
///     limit = 10,
/// });
/// Console.WriteLine(rows[0]!["name"]);
/// </code>
/// </summary>
public sealed class Orm : IDisposable
{
    private readonly Client _client;
    private IntPtr _schema;

    internal Orm(Client client, string schemaJson)
    {
        _client = client;
        _schema = Native.OrmSchemaNew(schemaJson);
        if (_schema == IntPtr.Zero)
        {
            throw new PowderException(Native.LastErrorString());
        }
    }

    /// <summary>Handle for one table's CRUD surface.</summary>
    public OrmTable Table(string name) => new(this, name);

    public void Dispose()
    {
        if (_schema != IntPtr.Zero)
        {
            Native.OrmSchemaFree(_schema);
            _schema = IntPtr.Zero;
        }
    }

    internal long Execute(string opJson)
    {
        long n = Native.OrmExecute(_client.Handle, SchemaHandle(), opJson);
        if (n < 0)
        {
            throw new PowderException(Native.LastErrorString());
        }
        return n;
    }

    internal JsonNode? FindJson(string opJson)
    {
        var ptr = Native.OrmFindJson(_client.Handle, SchemaHandle(), opJson, out nuint len);
        if (ptr == IntPtr.Zero)
        {
            throw new PowderException(Native.LastErrorString());
        }
        try
        {
            var bytes = new byte[(int)len];
            Marshal.Copy(ptr, bytes, 0, (int)len);
            return JsonNode.Parse(bytes);
        }
        finally
        {
            Native.FreeBuffer(ptr, len);
        }
    }

    private IntPtr SchemaHandle()
    {
        if (_schema == IntPtr.Zero)
        {
            throw new PowderException("orm is disposed");
        }
        return _schema;
    }
}

/// <summary>One table's unified CRUD surface.</summary>
public sealed class OrmTable
{
    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        // Keep anonymous-object property names exactly as written (`user_id`,
        // `AND`, `_count`, ...) — the op spec is case-sensitive.
        PropertyNamingPolicy = null,
    };

    private readonly Orm _orm;
    private readonly string _name;

    internal OrmTable(Orm orm, string name)
    {
        _orm = orm;
        _name = name;
    }

    /// <summary>Rows matching <paramref name="opts"/>; null opts = all rows.</summary>
    public JsonArray FindMany(object? opts = null) =>
        (JsonArray)_orm.FindJson(Op("findMany", opts))!;

    /// <summary>First matching row, or null.</summary>
    public JsonObject? FindFirst(object? opts = null) =>
        _orm.FindJson(Op("findFirst", opts)) as JsonObject;

    /// <summary>Every row.</summary>
    public JsonArray All() => FindMany();

    /// <summary>INSERT one row; missing (nullable) columns are omitted.</summary>
    public long Create(object data) => _orm.Execute(Op("create", new { data }));

    /// <summary>Bulk INSERT (chunked multi-row VALUES); every row must carry
    /// the same columns as the first.</summary>
    public long CreateMany(IEnumerable<object> rows) =>
        _orm.Execute(Op("createMany", new { rows }));

    /// <summary>UPDATE matching rows; returns the affected count.</summary>
    public long Update(object where, object data) =>
        _orm.Execute(Op("update", new { where, data }));

    /// <summary>DELETE matching rows. An empty where is rejected — use
    /// <see cref="DeleteAll"/>.</summary>
    public long Delete(object where) => _orm.Execute(Op("delete", new { where }));

    /// <summary>DELETE every row (explicit opt-in).</summary>
    public long DeleteAll() => _orm.Execute(Op("deleteAll", null));

    /// <summary>COUNT rows matching where (null counts everything).</summary>
    public long Count(object? where = null) =>
        _orm.Execute(Op("count", where == null ? null : new { where }));

    /// <summary>Whether at least one row matches.</summary>
    public bool Exists(object? where = null) =>
        FindFirst(where == null ? new { limit = 1 } : new { where, limit = 1 }) != null;

    /// <summary>SUM/AVG/MIN/MAX over one column; null when no rows match.</summary>
    public double? Aggregate(string fn, string column, object? where = null)
    {
        var v = _orm.FindJson(Op("aggregate",
            where == null ? new { fn, column } : new { fn, column, where }));
        return v == null ? null : v.GetValue<double>();
    }

    /// <summary>GROUP BY with aggregates (<c>by</c>, <c>count</c>, <c>sum</c>,
    /// <c>avg</c>, <c>min</c>, <c>max</c>, <c>having</c>, <c>orderBy</c>, ...);
    /// aggregates come back aliased <c>_count</c>, <c>_sum_&lt;col&gt;</c>, ....</summary>
    public JsonArray GroupBy(object opts) => (JsonArray)_orm.FindJson(Op("groupBy", opts))!;

    private string Op(string op, object? opts)
    {
        var node = opts == null
            ? new JsonObject()
            : (JsonObject)JsonSerializer.SerializeToNode(opts, JsonOpts)!;
        node["op"] = op;
        node["table"] = _name;
        return node.ToJsonString(JsonOpts);
    }
}
